# Test

## Output
Outputs:

PgBouncerIP=34.30.234.89
PgPublicIp=35.224.191.29

PgBouncer="pgbouncer-instance-e4bc2727f0"
db_connection_name="lunar-outlet-403221:us-central1:db-pgbouncer-0e45f4aa"
read_replica_connection_name=[
  "lunar-outlet-403221:us-east1:db-pgbouncer-0e45f4aa-replica-ha-0",
]

## Install PgBouncer SQL
The admin users of DB instance created by Terraform do not have password, set them at
https://console.cloud.google.com/sql/instances/db-pgbouncer-0e45f4aa/users?project=lunar-outlet-403221

    admin/admin
    postgres/postgres

    psql postgres://postgres:postgres@${PgPublicIp}:5432/testdb

Copy and paste [pgbouncer-install.sql](pgbouncer-install4.sql), this will create role `pgbouncer` and
    FUNCTION pgbouncer.get_auth(username TEXT)

## Test if role pgbouncer has SELECT privilege
on view `pg_shadow` and table `pg_authid`

This is b/c Terraform sets two database flags to "pgbouncer" in main.tf/database_flags, refer
https://cloud.google.com/sql/docs/postgres/users#setting_the_flags_for_the_pg_shadow_view_and_the_pg_authid_table
https://cloud.google.com/sql/docs/postgres/flags#list-flags-postgres

    psql postgres://pgbouncer:pgbouncer@${PgPublicIp}:5432/testdb

    SELECT usename, passwd FROM pg_shadow;
    SELECT rolname, rolpassword FROM pg_authid;

If not, there will be the following errors respectively:

        ERROR:  permission denied for view pg_shadow
        ERROR:  permission denied for table pg_authid

## Run Tests outside
    psql postgres://testuser:aPass@${PgBouncerIP}:6432/testdb
            psql: error: connection to server at "34.30.234.89", port 6432 failed: FATAL:  server login has been failing, try again later (server_login_retry)


## Debug in PgBouncer VM
    # SSH into ${PgBouncer}
    curl -s http://api.ipify.org/
    docker ps

CONTAINER ID   IMAGE                                               COMMAND                  CREATED          STATUS          PORTS                                                 NAMES
c97f6800d5f9   tpcorg/hammerdb:postgres                            "/bin/sh -c 'tail -f…"   59 minutes ago   Up 59 minutes                                                         hammerdb
9d48d4470fe3   edoburu/pgbouncer:latest                            "/entrypoint.sh /usr…"   59 minutes ago   Up 59 minutes   5432/tcp, 0.0.0.0:6432->6432/tcp, :::6432->6432/tcp   pgbouncer
53289694760a   gcr.io/cloud-sql-connectors/cloud-sql-proxy:2.1.2   "/cloud-sql-proxy lu…"   59 minutes ago   Up 59 minutes   0.0.0.0:5432->5432/tcp, :::5432->5432/tcp             cloudsql-proxy

    docker logs pgbouncer
            098 DEBUG C-0x7f126cfc3360: (nodb)/(nouser)@69.174.173.95:35576 P: got connection: 69.174.173.95:35576 -> 172.18.0.3:6432
            128 DEBUG C-0x7f126cfc3360: (nodb)/(nouser)@69.174.173.95:35576 got var: user=testuser
            128 DEBUG C-0x7f126cfc3360: (nodb)/(nouser)@69.174.173.95:35576 got var: database=testdb
            128 DEBUG C-0x7f126cfc3360: (nodb)/(nouser)@69.174.173.95:35576 using application_name: psql - 69.174.173.95:35576
            128 DEBUG C-0x7f126cfc3360: (nodb)/(nouser)@69.174.173.95:35576 got var: client_encoding=UTF8
            128 LOG   C-0x7f126cfc3360: testdb/(nouser)@69.174.173.95:35576 closing because: server login has been failing, try again later (server_login_retry) (age=0s)
            128 WARN  C-0x7f126cfc3360: testdb/(nouser)@69.174.173.95:35576 pooler error: server login has been failing, try again later (server_login_retry)
            144 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@(bad-af):0 dns_callback: inet4: 172.18.0.2:5432
            144 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 launching new connection to server
            145 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 S: connect ok
            145 LOG   S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 new connection to server (from 172.18.0.3:53688)
            161 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 calling login_answer
            161 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 S: req SASL
            161 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 S: SASL advertised mechanism: SCRAM-SHA-256
            161 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 SCRAM client-first-message = "n,,n=,r=I4n/BgUx1wSXO7j8XKdzlgD7"
            161 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 P: send SASLInitialResponse
            162 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 calling login_answer
            163 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 S: SASL cont
            163 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 SCRAM server-first-message = "r=I4n/BgUx1wSXO7j8XKdzlgD7u86SYdpkfSlU78IykHbrElr8,s=BqXYMuJ24NlFVNdVukaWOg==,i=4096"
            170 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 SCRAM client-final-message = "c=biws,r=I4n/BgUx1wSXO7j8XKdzlgD7u86SYdpkfSlU78IykHbrElr8,p=wnlJMyBTiUvHo+BWtzBRtpXCwbsqubiyCBwyXAPhvKM="
            170 DEBUG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 P: send SASLResponse
            172 WARNING server login failed: FATAL password authentication failed for user "pgbouncer"
            172 LOG S-0x7f126cf6e380: testdb/pgbouncer@172.18.0.2:5432 closing because: login failed (age=0s)
    CloudSQL log:
            [3149]: [1-1] db=testdb,user=pgbouncer FATAL:  password authentication failed for user "pgbouncer"

### Add user "pgbouncer" to userlist.txt
"pgbouncer" "md5be5544d3807b54dd0637f2439ecb03b9"

    cd /run/user
    stop_all_services.sh
    start_all_services.sh

Try again

            299 DEBUG C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 P: got connection: 69.174.173.95:35544 -> 172.20.0.3:6432
            329 DEBUG C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 got var: user=testuser
            329 DEBUG C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 got var: database=testdb
            329 DEBUG C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 using application_name: psql - 69.174.173.95:35544
            329 DEBUG C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 got var: client_encoding=UTF8
            330 DEBUG pktbuf_dynamic(128): 0x7f3e1921b270
            330 LOG   C-0x7f3e191f9360: (nodb)/(nouser)@69.174.173.95:35544 registered new auto-database: testdb
            330 DEBUG C-0x7f3e191f9360: testdb/(nouser)@69.174.173.95:35544 pause_client
            330 DEBUG zone_register(cloudsql-proxy)
            341 DEBUG launch_new_connection: already progress
            351 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@(bad-af):0 dns_callback: inet4: 172.20.0.2:5432
            351 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 launching new connection to server
            352 DEBUG launch_new_connection: already progress
            352 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 S: connect ok
            352 LOG   S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 new connection to server (from 172.20.0.3:34890)
            352 DEBUG launch_new_connection: already progress
            366 DEBUG pktbuf_dynamic(512): 0x7f3e191f4c00
            366 DEBUG launch_new_connection: already progress
            369 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 calling login_answer
            369 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 S: req SASL
            369 DEBUG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 S: SASL advertised mechanism: SCRAM-SHA-256
            369 ERROR S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 cannot do SCRAM authentication: wrong password type
            369 LOG S-0x7f3e191a4380: testdb/pgbouncer@172.20.0.2:5432 closing because: failed to answer authreq (age=0s)
            ......
            753 DEBUG launch_new_connection: last failed, not launching new connection yet, still waiting 2 s
            087 DEBUG launch_new_connection: last failed, not launching new connection yet, still waiting 1 s
            420 LOG C-0x7f3e191f9360: testdb/(nouser)@69.174.173.95:35544 closing because: query_wait_timeout (age=120s)
            420 WARNING C-0x7f3e191f9360: testdb/(nouser)@69.174.173.95:35544 pooler error: query_wait_timeout

### Set pgbouncer password to value from PG
    SELECT rolname, rolpassword FROM pg_authid WHERE rolname = 'pgbouncer';
              rolname  |                                                              rolpassword
            -----------+---------------------------------------------------------------------------------------------------------------------------------------
             pgbouncer | SCRAM-SHA-256$4096:BqXYMuJ24NlFVNdVukaWOg==$G9AmrEw5HKKOtF8T4plJhSnujGaOLZ8Aer1qj8EldA8=:gkPMX9oCX/1XX9KGXnNi9D4JbKz0IieU3Ao5iy4r+nQ=
            (1 row)

Try again

            18:33:26.854 NOISE S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 read pkt='R', len=24
            18:33:26.854 DEBUG S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 calling login_answer
            18:33:26.854 DEBUG S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 S: req SASL
            18:33:26.854 DEBUG S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 S: SASL advertised mechanism: SCRAM-SHA-256
            18:33:26.854 ERROR S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 cannot do SCRAM authentication: password is SCRAM secret but client authentication did not provide SCRAM keys
            18:33:26.854 LOG   S-0x7f5c724f6380: testdb/pgbouncer@172.21.0.2:5432 closing because: failed to answer authreq (age=0s)

Googled "pgbouncer cannot do SCRAM authentication: password is SCRAM secret but client authentication did not provide SCRAM keys",
Got https://github.com/pgbouncer/pgbouncer/issues/774
Try setting plain text password for role pgbouncer in userlist.txt

Try again,
        
        psql: error: connection to server at "34.30.234.89", port 6432 failed: FATAL:  bouncer config error

pgbouncer log:

        18:54:19.351 DEBUG pktbuf_dynamic(512): 0x7fd814f5cd40
        18:54:19.351 DEBUG pktbuf_free(0x7fd814f5cd40)
        18:54:19.358 LOG S-0x7fd814f0c380: testdb/pgbouncer@172.22.0.2:5432 closing because: error response from auth_query (age=0s)
        18:54:19.358 LOG C-0x7fd814f61360: testdb/(nouser)@69.174.173.95:35799 closing because: bouncer config error (age=0s)
        18:54:19.359 WARNING C-0x7fd814f61360: testdb/(nouser)@69.174.173.95:35799 pooler error: bouncer config error

CloudSQL log:

        [4836]: [1-1] db=testdb,user=pgbouncer ERROR:  permission denied for table pg_authid


    toolbox
    cd /media/root/run/user/
    ls
            clean_up.sh            configure-hammerdb.tcl  run_workload.sh   start_all_services.sh  terminate_active_connections.py
            configure-hammerdb.sh  pgbouncer.ini           run_workload.tcl  stop_all_services.sh   userlist.tx

    PgBouncer="pgbouncer-instance-9bf27e7161"
    db_connection_name="lunar-outlet-403221:us-central1:db-pgbouncer-32daac14"

    # List listening ports
    sudo lsof -nP -iTCP -sTCP:LISTEN

    apt-get update
    apt-get install postgresql-client

    psql postgres://testuser:aPass@${PgBouncer}:6432/testdb
            psql: error: FATAL:  SASL authentication failed

    psql -h $PgBouncer -p 6432 -U testuser testdb
            Password for user testuser: 
            psql: error: FATAL:  SASL authentication failed

    # Example PgBouncer's log
    docker logs pgbouncer

          DEBUG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 P: got connection: 10.128.0.2:34656 -> 172.18.0.3:6432
          DEBUG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 got var: user=testuser
          DEBUG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 got var: database=testdb
          DEBUG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 using application_name: psql - 10.128.0.2:34656
          DEBUG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 got var: client_encoding=SQL_ASCII
          LOG C-0x7f62ed59e360: (nodb)/(nouser)@10.128.0.2:34656 no such user: testuser
          LOG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 login attempt: db=testdb user=testuser tls=no
          DEBUG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 C: selected SASL mechanism: SCRAM-SHA-256
          DEBUG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 SCRAM client-first-message = "n,,n=,r=R1Yo3CNaqQPak01RBPUCDFdY"
          DEBUG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 SCRAM server-first-message = "r=R1Yo3CNaqQPak01RBPUCDFdY1/nlofaFPx761ToD2feAIR5Z,s=g/6n/DroJUM9wuxP2XdsqQ==,i=4096"
          DEBUG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 SCRAM client-final-message = "c=biws,r=R1Yo3CNaqQPak01RBPUCDFdY1/nlofaFPx761ToD2feAIR5Z,p=Tmm3E4o9iBqr61hakTERDU1Trm7rrc3aJ5yDuweNOSM="
          DEBUG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 SCRAM client-final-message-without-proof = "c=biws,r=R1Yo3CNaqQPak01RBPUCDFdY1/nlofaFPx761ToD2feAIR5Z"
          ERROR C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 password authentication failed
          LOG C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 closing because: SASL authentication failed (age=0s)
          WARNING C-0x7f62ed59e360: (nodb)/testuser@10.128.0.2:34656 pooler error: SASL authentication failed

          "password authentication failed" is expected, b/c we haven't create the auth id query function in the DB yet.

    # Example cloudsql-proxy's log
    docker logs cloudsql-proxy

          2023/11/21 22:40:47 Authorizing with Application Default Credentials
          2023/11/21 22:40:48 [lunar-outlet-403221:us-central1:db-pgbouncer-32daac14] Listening on [::]:5432
          2023/11/21 22:40:48 [lunar-outlet-403221:us-east1:db-pgbouncer-32daac14-replica-ha-0] Listening on [::]:5433
          2023/11/21 22:40:48 The proxy has started successfully and is ready for new connections!

## Test
    Error in PG log:
        user=pgbouncer ERROR:  permission denied for table pg_authid

    \dp pg_catalog.pg_authid

## Add Access to the pg_shadow view and the pg_authid table
[>](https://cloud.google.com/sql/docs/postgres/users#pg_shadow)

cloudsql.pg_authid_select_role
